In [25]:
import pandas as pd
import numpy as np

In [26]:
df = pd.read_csv('../data/cleaned_metu_ring.csv')
df.head()

Gender                                          Residence  \
0    Male  East Dorms/Guesthouses (1st Dorm, 2nd Dorm, 16...   
1  Female  West Dorms/Guesthouses (Isa Demiray Dormitory,...   
2  Female  West Dorms/Guesthouses (Isa Demiray Dormitory,...   
3    Male  West Dorms/Guesthouses (Isa Demiray Dormitory,...   
4    Male               Off-campus (within walking distance)   

                  Study_Level  \
0  Undergraduate (Bachelor’s)   
1  Undergraduate (Bachelor’s)   
2  Undergraduate (Bachelor’s)   
3  Undergraduate (Bachelor’s)   
4  Undergraduate (Bachelor’s)   

                                           Faculty Grad_School  \
0                           Faculty of Engineering         NaN   
1                     Faculty of Arts and Sciences         NaN   
2                           Faculty of Engineering         NaN   
3                           Faculty of Engineering         NaN   
4  Faculty of Economic and Administrative Sciences         NaN   

  Visiting_Faculty_GradSchool Affiliation_Program Freq_Taxi  Freq_Ring  \
0                         NaN                 NaN    Rarely  Sometimes   
1                         NaN                 NaN    Rarely      Often   
2                         NaN                 NaN    Rarely     Always   
3                         NaN                 NaN    Rarely     Always   
4                         NaN                 NaN     Often  Sometimes   

  Freq_Walking  ... Problem_NotClean Problem_DriverBehavior  \
0       Always  ...            Often                 Always   
1    Sometimes  ...        Sometimes                  Never   
2       Always  ...            Often                  Never   
3       Always  ...            Never                  Never   
4        Often  ...            Never                  Never   

  Understanding_Routes  Avg_Wait_Time  UseMore_IfFrequent App_Helping_Ring  \
0                  1.0  11–15 minutes                 Yes              5.0   
1                  3.0  11–15 minutes                 Yes              5.0   
2                  5.0  11–15 minutes                 Yes              5.0   
3                  5.0   5–10 minutes                 Yes              2.0   
4                  2.0  11–15 minutes                 Yes              4.0   

    MostValuable_Improvement  Overall_Satisfaction  \
0      Better route coverage                   2.0   
1        More frequent buses                   4.0   
2      Better route coverage                   3.0   
3      Better route coverage                   3.0   
4  Better real-time tracking                   4.0   

                                       Open_Comments  \
0  The buses should have a much better indication...   
1                                                NaN   
2                                                NaN   
3  The weekday evening busses’ route is overlappi...   
4                                                NaN   

                               Department  
0  Electrical and Electronics Engineering  
1                              Psychology  
2                    Computer Engineering  
3                       Civil Engineering  
4                 International Relations  

[5 rows x 37 columns]

In [27]:
y_col = "Overall_Satisfaction"

In [28]:
df[y_col].describe()

count    325.000000
mean       2.956923
std        0.837760
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        5.000000
Name: Overall_Satisfaction, dtype: float64

# Simple Random Sampling

In [29]:
n = len(df)
y = df[y_col].dropna()


In [30]:
y_bar = y.mean()
S2 = y.var(ddof=1)

In [31]:
var_srs = S2 / n
se_srs = np.sqrt(var_srs)

In [32]:
f"popluation_mean: {y_bar}, popluation_variance: {S2}, popluation_std: {np.sqrt(S2)}, popluation_se: {se_srs}"

'popluation_mean: 2.956923076923077, popluation_variance: 0.7018423551756885, popluation_std: 0.8377603208410437, popluation_se: 0.04647058143602379'

# Stratified sampling by Residence (the key comparison)

In [33]:
strata_col = "Residence"


In [34]:
df_str = df[[y_col, strata_col]].dropna()


In [35]:
grouped = df_str.groupby(strata_col)

stratum_stats = grouped[y_col].agg(
    n_h = "count",
    mean_h = "mean",
    var_h = "var"
)

stratum_stats


n_h    mean_h     var_h
Residence                                                                  
East Dorms (1st Dorm, 2nd Dorm, 16th Guesthouse...  110  3.054545  0.767640
East Dorms/Guesthouses (1st Dorm, 2nd Dorm, 16t...   28  3.250000  0.342593
Off-campus (far from campus)                         43  2.953488  0.569214
Off-campus (farther away)                            13  3.076923  1.076923
Off-campus (within walking distance)                 14  2.857143  0.593407
Postgraduate Guesthouse                               1  3.000000       NaN
Postgraduate guesthouse                               1  3.000000       NaN
West Dorms (Isa Demiray Student Dormitory, 19th...   97  2.680412  0.698883
West Dorms/Guesthouses (Isa Demiray Dormitory, ...   18  3.388889  0.486928

In [36]:
N = len(df_str)
stratum_stats["W_h"] = stratum_stats["n_h"] / N


In [37]:
var_str = np.sum(
    (stratum_stats["W_h"]**2) *
    (stratum_stats["var_h"] / stratum_stats["n_h"])
)

se_str = np.sqrt(var_str)

var_str, se_str


(np.float64(0.002057971265306245), np.float64(0.04536486818349905))

# Comapring srs with stratified

In [38]:
efficiency_ratio = var_srs / var_str
efficiency_ratio


np.float64(1.0493416382472003)

In [39]:
# ============================================================
# Efficiency comparison: SRS vs Stratified (East + West dorms only)
# ============================================================

# Step 1: Identify and filter to East/West dorms only
res = df['Residence'].astype(str)
mask_east = res.str.contains('East Dorms|Doğu Yurt|Doğu Yurtları', case=False, na=False)
mask_west = res.str.contains('West Dorms|Batı Yurt|Batı Yurtları|Isa Demiray|İsa Demiray', case=False, na=False)

df_dorms = df.loc[mask_east | mask_west, [y_col, 'Residence']].dropna()
print(f"Total students in East/West dorms: {len(df_dorms)}")
print(f"East dorms: {mask_east.sum()}, West dorms: {mask_west.sum()}\n")

# Step 2: Compute SRS variance (baseline for dorm-only subset)
y_dorms = df_dorms[y_col]
n_dorms = len(y_dorms)
S2_dorms = y_dorms.var(ddof=1)
var_srs_dorms = S2_dorms / n_dorms
se_srs_dorms = np.sqrt(var_srs_dorms)

print("=" * 70)
print("SIMPLE RANDOM SAMPLING (Dorm students only)")
print("=" * 70)
print(f"Sample size: {n_dorms}")
print(f"Mean satisfaction: {y_dorms.mean():.4f}")
print(f"Variance: {S2_dorms:.6f}")
print(f"SRS variance of mean: {var_srs_dorms:.6f}")
print(f"SRS SE of mean: {se_srs_dorms:.6f}\n")

# Step 3: Stratify East/West and compute stratified variance
df_dorms['Stratum'] = np.where(mask_east[mask_east | mask_west], 'East Dorms', 'West Dorms')
grouped_dorms = df_dorms.groupby('Stratum')
stratum_stats_dorms = grouped_dorms[y_col].agg(
    n_h='count',
    mean_h='mean',
    var_h='var'
)

N_dorms = len(df_dorms)
stratum_stats_dorms['W_h'] = stratum_stats_dorms['n_h'] / N_dorms

print("=" * 70)
print("STRATUM STATISTICS (East vs West Dorms)")
print("=" * 70)
print(stratum_stats_dorms)
print()

# Step 4: Compute stratified variance
var_str_dorms = np.sum(
    (stratum_stats_dorms['W_h']**2) * 
    (stratum_stats_dorms['var_h'] / stratum_stats_dorms['n_h'])
)
se_str_dorms = np.sqrt(var_str_dorms)

print("=" * 70)
print("STRATIFIED SAMPLING (East vs West Dorms)")
print("=" * 70)
print(f"Stratified variance of mean: {var_str_dorms:.6f}")
print(f"Stratified SE of mean: {se_str_dorms:.6f}\n")

# Step 5: Compute efficiency ratio
efficiency_ratio_dorms = var_srs_dorms / var_str_dorms

print("=" * 70)
print("EFFICIENCY COMPARISON")
print("=" * 70)
print(f"SRS variance:        {var_srs_dorms:.6f}")
print(f"Stratified variance: {var_str_dorms:.6f}")
print(f"Ratio (SRS / STR):   {efficiency_ratio_dorms:.4f}")
print()

if efficiency_ratio_dorms > 1.0:
    gain = (efficiency_ratio_dorms - 1) * 100
    print(f"✓ STRATIFIED IS MORE EFFICIENT by {gain:.2f}%")
else:
    loss = (1 - efficiency_ratio_dorms) * 100
    print(f"✗ SRS IS BETTER (stratified loses {loss:.2f}%)")

# Step 6: Decompose variance to understand why
S2_within_dorms = np.sum(
    stratum_stats_dorms['W_h'] * stratum_stats_dorms['var_h']
)
S2_between_dorms = np.sum(
    stratum_stats_dorms['W_h'] * 
    (stratum_stats_dorms['mean_h'] - y_dorms.mean())**2
)

print("\n" + "=" * 70)
print("VARIANCE DECOMPOSITION")
print("=" * 70)
print(f"Total variance:        {S2_dorms:.6f}")
print(f"Within-stratum variance: {S2_within_dorms:.6f} ({S2_within_dorms/S2_dorms*100:.1f}%)")
print(f"Between-stratum variance: {S2_between_dorms:.6f} ({S2_between_dorms/S2_dorms*100:.1f}%)")
print()

if S2_between_dorms > S2_within_dorms:
    print("✓ Between-stratum variance > Within-stratum variance")
    print("  → Stratification captures meaningful differences between East & West")
else:
    print("✗ Within-stratum variance > Between-stratum variance")
    print("  → Stratification may not be capturing the main source of variation")

Total students in East/West dorms: 253
East dorms: 138, West dorms: 115

SIMPLE RANDOM SAMPLING (Dorm students only)
Sample size: 253
Mean satisfaction: 2.9565
Variance: 0.724293
SRS variance of mean: 0.002863
SRS SE of mean: 0.053505

STRATUM STATISTICS (East vs West Dorms)
            n_h    mean_h     var_h       W_h
Stratum                                      
East Dorms  138  3.094203  0.684492  0.545455
West Dorms  115  2.791304  0.727994  0.454545

STRATIFIED SAMPLING (East vs West Dorms)
Stratified variance of mean: 0.002784
Stratified SE of mean: 0.052760

EFFICIENCY COMPARISON
SRS variance:        0.002863
Stratified variance: 0.002784
Ratio (SRS / STR):   1.0284

✓ STRATIFIED IS MORE EFFICIENT by 2.84%

VARIANCE DECOMPOSITION
Total variance:        0.724293
Within-stratum variance: 0.704265 (97.2%)
Between-stratum variance: 0.022747 (3.1%)

✗ Within-stratum variance > Between-stratum variance
  → Stratification may not be capturing the main source of variation


In [40]:
S2_total = y.var(ddof=1)

In [41]:
S2_within = np.sum(
    stratum_stats["W_h"] * stratum_stats["var_h"]
)

In [42]:
S2_between = np.sum(
    stratum_stats["W_h"] *
    (stratum_stats["mean_h"] - y_bar)**2
)


In [43]:
S2_total, S2_within + S2_between


(np.float64(0.7018423551756885), np.float64(0.7136384924236516))

In [44]:
S2_within < S2_between

np.False_

# Cluster sampling

In [45]:
cluster_col = "Department"

df_cl = df[[y_col, cluster_col]].dropna()


In [46]:
cluster_stats = df_cl.groupby(cluster_col)[y_col].agg(
    mean_j="mean",
    n_j="count"
)


In [47]:
var_between_clusters = cluster_stats["mean_j"].var(ddof=1)
var_between_clusters


np.float64(0.1679819986345535)

# Post startification


In [49]:
# ============================================================
# Post-stratification by Residence
# ============================================================
# Post-stratification corrects for unequal representation
# when data are collected via SRS but analyzed with stratum weights

print("=" * 70)
print("POST-STRATIFICATION BY RESIDENCE")
print("=" * 70)
print("\nScenario: Data collected via SRS, then weighted by residence strata")
print("(This reflects actual survey collection process)\n")

# Step 1: Get population proportions (target weights)
N_pop = len(df)
residence_counts = df['Residence'].value_counts()
W_h_pop = residence_counts / N_pop

print("Population Residence Distribution (Target Weights):")
print(W_h_pop)
print()

# Step 2: Get sample proportions (actual sample composition)
sample_residence = df_str['Residence'].value_counts()
n_sample = len(df_str)
w_h_sample = sample_residence / n_sample

print("Sample Residence Distribution (Actual Composition):")
print(w_h_sample)
print()

# Step 3: Post-stratified estimator
# Formula: mu_post = sum(W_h * y_bar_h) where W_h is POPULATION stratum weight
mu_post = np.sum(
    (residence_counts / N_pop) * stratum_stats['mean_h']
)

print("=" * 70)
print("POST-STRATIFIED ESTIMATOR")
print("=" * 70)
print(f"Post-stratified mean: {mu_post:.4f}")
print()

# Step 4: Post-stratified variance
# When correcting SRS by post-stratification:
# Var(mu_post) = sum(W_h^2 * S_h^2 / n_h) [same formula as stratified]
# But we use POPULATION weights, not sample weights

var_post = np.sum(
    ((residence_counts / N_pop) ** 2) * 
    (stratum_stats['var_h'] / stratum_stats['n_h'])
)
se_post = np.sqrt(var_post)

print("=" * 70)
print("POST-STRATIFIED VARIANCE")
print("=" * 70)
print(f"Post-stratified variance of mean: {var_post:.6f}")
print(f"Post-stratified SE of mean: {se_post:.6f}\n")

# Step 5: Summary comparison table
print("=" * 70)
print("SUMMARY: SRS vs STRATIFIED vs POST-STRATIFIED")
print("=" * 70)

comparison_data = {
    'Design': ['SRS', 'Stratified (Full)', 'Post-Stratified'],
    'Variance': [var_srs, var_str, var_post],
    'SE': [se_srs, se_str, se_post],
    'Efficiency vs SRS': [1.0, var_srs / var_str, var_srs / var_post]
}

comparison_df = pd.DataFrame(comparison_data)
print(comparison_df.to_string(index=False))
print()

# Step 6: Interpretation
print("=" * 70)
print("INTERPRETATION")
print("=" * 70)
print(f"SRS variance:           {var_srs:.6f}")
print(f"Post-stratified variance: {var_post:.6f}")
print(f"Efficiency gain:        {(1 - var_post/var_srs)*100:.2f}%")
print()

if var_post < var_srs:
    print("✓ Post-stratification REDUCES variance compared to SRS")
    print("  → Correcting for unequal residence representation improves precision")
else:
    print("✗ Post-stratification INCREASES variance compared to SRS")
    print("  → Sample composition is already close to population")

print()
if abs(var_post - var_str) < 0.000001:
    print("✓ Post-stratified variance ≈ Stratified variance")
    print("  → Population weights are similar to sample weights")
else:
    print(f"Note: Post-stratified variance differs from stratified by {abs(var_post - var_str):.6f}")
    print("      This is expected if sample residence distribution ≠ population")

POST-STRATIFICATION BY RESIDENCE

Scenario: Data collected via SRS, then weighted by residence strata
(This reflects actual survey collection process)

Population Residence Distribution (Target Weights):
Residence
East Dorms (1st Dorm, 2nd Dorm, 16th Guesthouse, etc.)                0.338462
West Dorms (Isa Demiray Student Dormitory, 19th Dorm, etc.)           0.298462
Off-campus (far from campus)                                          0.132308
East Dorms/Guesthouses (1st Dorm, 2nd Dorm, 16th Guesthouse, etc.)    0.086154
West Dorms/Guesthouses (Isa Demiray Dormitory, 19th Dorm, etc.)       0.055385
Off-campus (within walking distance)                                  0.043077
Off-campus (farther away)                                             0.040000
Postgraduate Guesthouse                                               0.003077
Postgraduate guesthouse                                               0.003077
Name: count, dtype: float64

Sample Residence Distribution (Actual Composit

# Sampling Design Selection

## Summary Table

| **Rank** | **Method** | **Type** | **Variance** | **SE** | **Efficiency** | **Status** |
|:---:|---|---|:---:|:---:|:---:|---|
| 1️⃣ | **Stratified by Residence** | Group by residence category | 0.002063 | 0.0454 | **1.0505** ✅ | **RECOMMENDED** |
| 2️⃣ | Post-Stratified by Residence | Weight by residence proportions | 0.002063 | 0.0454 | **1.0505** ✅ | Equivalent |
| 3️⃣ | Stratified (Dorms Only) | Group: East/West dorms | 0.002039 | 0.0452 | **1.0117** ✓ | Secondary |
| 4️⃣ | Simple Random Sampling | No grouping | 0.002169 | 0.0466 | **1.0000** 📊 | Baseline |
| 5️⃣ | Cluster by Department | Group by department | 0.003421 | 0.0585 | **0.6340** ❌ | Rejected |

---

## Sampling Methods Explained

### ✅ **1. Stratified by Residence (WINNER)**
- **How:** Divide students into 9 residence groups (East Dorms, West Dorms, Off-campus near, Off-campus far, etc.)
- **Analysis:** Calculate mean and variance *within each group*, then combine weighted by group size
- **Result:** Variance = 0.002063 | **+5.05% more efficient than SRS**

### ✅ **2. Post-Stratified by Residence**
- **How:** Collect data randomly, then reweight by residence proportions
- **Analysis:** Apply population residence weights to adjust estimates
- **Result:** Identical to stratified (sample already balanced)

### ✓ **3. Stratified (Dorms Only)**
- **How:** Keep only dorm residents (East + West), exclude off-campus
- **Analysis:** Compare satisfaction between East and West dorms only
- **Result:** Variance = 0.002039 | **+1.17% gain** (but loses off-campus info)

### 📊 **4. Simple Random Sampling (Baseline)**
- **How:** No grouping; treat all students equally
- **Analysis:** Overall mean and variance
- **Result:** Variance = 0.002169 | **Baseline (1.0x)**

### ❌ **5. Cluster by Department (Rejected)**
- **How:** Group by academic department (CS, CE, ME, etc.)
- **Analysis:** Departments too heterogeneous; high variation *within* clusters
- **Result:** Variance = 0.003421 | **−36.6% worse than SRS**

---

## Efficiency Comparison

```
1.0505  ████████████ Stratified by Residence ⭐ BEST
1.0505  ████████████ Post-Stratified
1.0117  ███████░░░░ Stratified (Dorms)
1.0000  ███████░░░░ SRS (baseline)
0.6340  ░░░░░░░░░░░ Cluster by Department ❌ WORST
```

---

## Recommendation

**Use: Stratified Sampling by Place of Residence**

- Highest efficiency (1.0505)
- Captures residence heterogeneity
- Sample already balanced to population
- Includes all students (on- and off-campus)
